### ISSUE TITLE
KERNEL_SECURITY_CHECK FAILURE MMap Driver

### ISSUE THREAD

**Issue by Screeder** [DarthTon/Blackbone#167]
Hello,

I have a problem with the usage of a manual mapped driver and hopefully someone can help me.

I use the Blackbone driver to map a second blackbone driver and don't initialize anything in DriverEntry since both values are NULL. Now if I try to use a method like BBInitDynamicData I get a bluescreen with the bug check name in the header. Especially in method BBInitDynamicData the RtlGetVersion method is causing a bluescreen. Many other methods are causing the same problem. I don't know how a manual mapped driver would be useful this way...

I also tried to use PsSetCreateProcessNotifyRoutine but I get the error ACCESS_DENIED. I have already removed wipe header for manual mapping and added /INTEGRITYCHECK for the second driver.

Do you know how can I fix these problems?

Thanks in advance

**Comment by koemeet** [comment]
You should provide a crash dump and some info about your environment, windows version etc.

**Comment by DarthTon** [comment]
Probably this is normal for not having proper backing module structures, but I'll look into it. Dump will help indeed.

**Comment by Screeder** [comment]
VMWare Windows 10 x64 (Version 1703, Build 15063.483)
Driver compiled with VS 15, Testapp compiled with VS 17
Source based on commit: https://github.com/DarthTon/Blackbone/commit/ab1ab817a9ae78537bc44090f13bf16ac77920db

You can fix the bsod by set /GS to /GS- .
Also if you pass your DriverEntry from your normal driver to the manually mapped driver it can initialize PsLoadedModuleList.
After that almost any function works but PsSetCreateProcessNotifyRoutine not because I think it needs a valid DriverEntry for the manually mapped driver section.


[Dump](https://github.com/DarthTon/Blackbone/files/1167268/072217-43859-01.zip)

**Comment by staticinvocation** [comment]
You probably have a buffer overrun somewhere in your code.  You should probably fix that issue instead of treating the symptom by setting /GS-, as this could cause you even more issues later on.

Also, `PsSetCreateProcessNotifyRoutine` will check your callback routine address to make sure it is located in a module in `PsLoadedModuleList`.  Even if you manually map in a callback, this structure is scanned by Patchguard and you will bugcheck.  In short- there is no way to use `PsSetCreateProcessNotifyRoutine` in a manually-mapped driver without mapping the callback manually and disabling patchguard.

**Comment by Screeder** [comment]
So then RtlGetVersion causes an buffer overrun itself? Because that is the first "real" call and this causes the bug check. I had commented out everything else and https://github.com/DarthTon/Blackbone/blob/master/src/BlackBoneDrv/BlackBoneDrv.c#L35 after calling this bb function followed by 
https://github.com/DarthTon/Blackbone/blob/master/src/BlackBoneDrv/BlackBoneDrv.c#L242 causes the bugsplat. Nothing more gets executed.

But nice to know that Patchguard is checking it. Therefore I couldn't redirect the call of a ProcessNotifyRoutine to my manually mapped driver from another driver...

**Comment by staticinvocation** [comment]
So you're trying to map Blackbone with Blackbone?  If so, that's not going to work, because Blackbone is dependent on a driver object.

**Comment by Screeder** [comment]
Yes I am using the Blackbone driver to manual map a second Blackbone driver as I stated in the main post. Also in my second post I said that you can use most of the Blackbone driver functions if you pass the DriverEntry from the first Blackbone driver to the manually mapped one. The only problem is that you bug check if you use PsSetCreateProcessNotifyRoutine.

So my solution is now that I pass the DriverEntry to the manually mapped driver of the first driver which injects the manually mapped driver and I won't let the manually mapped driver execute the PsSetCreateProcessNotifyRoutine method (disable Gs too). I am able with this workaround to use any Blackbone method in my manually mapped Blackbone driver.
